share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorOffsetsManager.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.coordinator.share;
+
+import org.apache.kafka.server.share.SharePartitionKey;
+import org.apache.kafka.timeline.SnapshotRegistry;
+import org.apache.kafka.timeline.TimelineHashMap;
+import org.apache.kafka.timeline.TimelineLong;
+
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Util class to track the offsets written into the internal topic
+ * per share partition key.
+ * It calculates the minimum offset globally up to which the records
+ * in the internal partition are redundant i.e. they have been overridden
+ * by newer records.
+ */
+public class ShareCoordinatorOffsetsManager {
+
+    // Map to store share partition key => current partition offset
+    // being written.
+    private final TimelineHashMap<SharePartitionKey, Long> offsets;
+
+    // Minimum offset representing the smallest necessary offset (non-redundant)
+    // across the internal partition.
+    // We are using timeline object here because the offsets which are passed into
+    // updateState might not be committed yet. In case of retry, these offsets would
+    // be invalidated via the snapshot registry. Hence, using timeline hashmaps
+    // the values would automatically revert in accordance with the last committed offset.
+    private final TimelineLong minOffset;
+    private final TimelineLong redundantOffset;
```

Comment:
Hmm, `minOffset` is defined as the smallest necessary offset. This seems to to be the same as `redundantOffset`?

Reply:
Part of https://github.com/apache/kafka/pull/18014#discussion_r1876665274

Reply:
Ok. So `redundantOffset` is the same as `minOffset` but is reset after it's exposed to the caller. It would be useful to document this. Also, I am wondering if it's clearer to maintain just minOffset and a flag that indicates whether it has been exposed or not.

Reply:
We could do that

Reply:
@junrao incorporated change